docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,212 @@
+    - Watches `Bastion` custom resources that are created by the `gardenlet` in the seed
+    - Controller reads `cloudprovider` credentials from seed-shoot namespace
+    - Deploy infrastructure resources
+        - Bastion VM. User data similar to https://github.com/gardener/gardenctl/blob/1e3e5fa1d5603e2161f45046ba7c6b5b4107369e/pkg/cmd/ssh.go#L160-L171. Writes `spec.sshPublicKey` into `authorized_keys` file.
+        - attaches public IP, creates security group, firewall rules, etc.
+    - Updates status of `Bastion` resource:
+        - With bastion IP under `status.ingress.ip` or hostname under `status.ingress.hostname`
+        - Updates the `status.lastOperation` with the status of the last reconcile operation
+6. `gardenlet`
+    - Syncs back the `status.ingress` and `status.conditions` of the `Bastion` resource in the seed to the garden cluster in case it changed
+7. `gardenctl`
+    - initiates `ssh` session once `status.conditions['BastionReady']` is true of the `Bastion` resource in the garden cluster
+        - locates private `ssh` key matching `spec["sshPublicKey"]` which was configured beforehand by the user
+        - reads bastion IP (`status.ingress.ip`) or hostname (`status.ingress.hostname`)
+        - reads the private key from the `ssh` key pair for the shoot node
+        - opens `ssh` connection to the bastion and from there to the respective shoot node
+    - runs heartbeat in parallel as long as the `ssh` session is open by annotating the `Bastion` resource with `operations.gardener.cloud/operation: keepalive`
+8. `GCM`:
+    - Once `status.expirationTimestamp` is reached, the `Bastion` will be marked for deletion
+9. `gardenlet`:
+    - Once the `Bastion` resource in the garden cluster is marked for deletion, it marks the `Bastion` resource in the seed for deletion
+10. Gardener extension provider <infra> / Bastion Controller on Seed:
+    - all created resources will be cleaned up
+    - On succes, removes finalizer on `Bastion` resource in seed
+11. `gardenlet`:
+    - removes finalizer on `Bastion` resource in garden cluster
+
+### Resource Example
+
+`Bastion` resource in the garden cluster
+```yaml
+apiVersion: operations.gardener.cloud/v1alpha1
+kind: Bastion
+metadata:
+  generateName: cli-
+  name: cli-abcdef
+  namespace: garden-myproject
+  annotations:
+    operations.gardener.cloud/created-by: foo # set by the mutating webhook
+    # operations.gardener.cloud/operation: keepalive # this annotation is removed by the GAPI and the last-heartbeat timestamp and the status.expirationTimestamp will be updated accordingly
+spec:
+  shootRef: # namespace cannot be set / it's the same as .metadata.namespace
+    name: my-cluster
+
+  sshPublicKey: c3NoLXJzYSAuLi4K # public `ssh` key of the user
+
+  ingress:
+  - ipBlock:
+      cidr: 1.2.3.4/32 # public IP of the user. CIDR is a string representing the IP Block. Valid examples are "192.168.1.1/24" or "2001:db9::/64"
```

Comment:
Probably this has already been discussed in some other comments but I couldn't read through all of them, still this sounds a bit strange to me...

How do we prevent the user from specifying "too broad" network ranges? And how do we define "too broad"?
I mean, surely the user shouldn't be allowed to specify "0.0.0.0/0", but is "1.2.0.0/16" already "narrow" enough? 
Specifying a CIDR like this sounds purely like ticking a checkbox to satisfy some security requirement, but doesn't seem to provide a real security benefit or might even cause some additional hassle (e.g. what if the user's public IP changes every few hours (not uncommon for some cable ISPs at least) and they can't predict, which CIDR is actually used).

Reply:
No it hasn't been discussed before

Do we want to prevent it? If the range is too broad it will most likely fire an alert in case it is checked by compliance / security tools and/or is automatically removed by such solutions. But yes we could, if we see the need for it. The maximum allowed range could also be configurable.

Well yes it's ticking a security requirement, but it is also adding some level of security if the range is narrow enough. Your client (e.g. `gardenctl`) could help you in case you phase issues in case your public IP changes often or if you switch off or on your VPN. The client just needs to update `spec.ingress[].ipBlock` section of the `Bastion` resource and after it is reconciled it should work again - at least that's the plan :)

Reply:
> Do we want to prevent it? If the range is too broad it will most likely fire an alert in case it is checked by compliance / security tools and/or is automatically removed by such solutions. But yes we could, if we see the need for it. The maximum allowed range could also be configurable.

Ok, so what was the reason then for going from specifying a client IP to a client IP cidr? (sorry, can't find it quickly in all the comments)
I don't like such half-measures (it is supposed to increase security but easily circumvented by the user in which case it will not really make it safer to have the restriction), but maybe that's just me.

> Your client (e.g. gardenctl) could help you in case you phase issues in case your public IP changes often or if you switch off or on your VPN. The client just needs to update spec.ingress[].ipBlock section of the Bastion resource and after it is reconciled it should work again - at least that's the plan :)

Ok, that's good then. I wasn't aware, that these fields are also supposed to be udpated and reconciled during an ongoing SSH session.

Reply:
Sorry for going into the discussion here again, originally I wanted to focus on the gardener-related/general aspects.
I will resolve this, as you are now aware of the concerns, but security is outside of my focus. I trust, you will find a good solution :)

Reply:
Do we have to care explicitly about it? From API point of view, it's in the user's responsibility to properly configure the ingress CIDRs (like in `NetworkPolicy`s). From user's (= gardenctl, most prominently) point of view, we simply take the public IP and add it here, nothing else.

Reply:
> Ok, so what was the reason then for going from specifying a client IP to a client IP cidr? (sorry, can't find it quickly in all the comments)

This was proposed by @rfranzke:
> rfranzke 10 days 
Again, I would prefer this to be similar to the ingress section of NetworkPolicys.

source: https://github.com/gardener/gardener/pull/3802#discussion_r607458539

Reply:
Yep, fine for me then.